### ISSUE TITLE
issues working with redshift

### ISSUE THREAD

**Issue by ofajardo** [machow/siuba#232]
Hi

I have recently started using both siuba and redshift and I am having some trouble I would like to report. First of all, I would like to congratulate you for the excellent work you are doing with siuba, I hope it keeps growing!

I am using Python 3.7 on ubuntu 18.04 running on kubernetes. I can give you more details about the setup if needed (I am also new to this setup).

Now, to the issues:

installing siuba 0.0.20 I get sqlalchemy == 1.3.16. I also install sqlalchemy-redshift 0.7.7.

When I do something like this:

```

connect_string = f'redshift+psycopg2://{db_user}:{db_password}@{db_host}:{db_port}/{db_name}'
cohort = (LazyTable(connect_string, "somedatabase.sometable") >>
                select("someid", 'colA', 'colB', "colC") >>
                left_join(_, LazyTable(connect_string, "somedatabase.anothertable") >>
                       select("someid","colD", "colE"),
                    on="someid") >>
              collect())
```

I get an error:

```
TypeError: _get_column_info() missing 1 required positional argument: 'generated'
```

full stack trace: 
[first_stack_trace.txt](https://github.com/machow/siuba/files/4623222/first_stack_trace.txt)

Then, I downgrade sqlalchemy to sqlalchemy == 1.3.15. The previous error disappears. 

But later I am doing something like this:

```
varname = "ColA"
df = ( LazyTable(connect_string, "somedatabase.sometable") >>
               filter( _.someid in alistofids) >>
               select("someid", varname ) >>
               rename(activity_date = varname) >>
               collect())
```

and I get this error:

```
/opt/conda/lib/python3.7/site-packages/sqlalchemy_redshift/dialect.py:761: SADeprecationWarning: The Engine.contextual_connect() and Connection.contextual_connect() methods are deprecated.  This method is an artifact of the threadlocal engine strategy which is also to be deprecated.   For explicit connections from an Engine, use the Engine.connect() method.

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-9-ad6152b6003d> in <module>
     14                sfilter( _.patientid in mm_pats) >>
     15                select("patientid", varname ) >>
---> 16                rename(activity_date = varname) >>
     17                collect())
     18         df.activity_date = pd.to_datetime(df.activity_date)

/opt/conda/lib/python3.7/site-packages/siuba/siu.py in __bool__(self)
    699 
    700     def __bool__(self):
--> 701         raise TypeError("Symbolic objects can not be converted to True/False, or used "
    702                         "with these keywords: not, and, or.")
    703 

TypeError: Symbolic objects can not be converted to True/False, or used with these keywords: not, and, or.
```

finally, I downgrade to siuba 0.0.17 (because 0.0.19 and 0.018 produce the same error) and now the error is gone, although I still get this warning:

```
/opt/conda/lib/python3.7/site-packages/sqlalchemy_redshift/dialect.py:761: SADeprecationWarning: The Engine.contextual_connect() and Connection.contextual_connect() methods are deprecated.  This method is an artifact of the threadlocal engine strategy which is also to be deprecated.   For explicit connections from an Engine, use the Engine.connect() method.
  with connection.contextual_connect() as cc:
```

it would be great if these issues could be repaired so that people could use the latest versions of the packages involved. 
I can test if needed. 

thanks in advance

**Comment by machow** [comment]
Hey @ofajardo, thanks for raising this issue--I think the problem occurs in this piece of code...

```python
sfilter( _.patientid in mm_pats)
```

There are two potential issues here.

### The `in` keyword likely doesn't do what you want in pandas

A kind of surprising thing is that, in pandas, `1 in some_column` checks whether `1` is in the some_column's **index, not its values**.

```python
# False
'a' in pd.Series(['a', 'b']) 
```


###  `_` can't work with python's `in` keyword

For example, the code below raises an error.

```python
from siuba.siu import _

_.a in ['x', 'y', 'z']
```

 It doesn't raise in previous siuba versions, but this is a bug! This [section of the docs](https://siuba.readthedocs.io/en/latest/guide_programming.html#Caveats) (esp the third point) explains some limitations of `_`.

Generally, `_` supports nearly all expressions!

## Solution

Can you try using something like the code below, and let me know if it fixes the issue? The key is the `.isin` method, which in pandas checks the values rather than index.

```python
from siuba.data import cars_sql
from siuba import filter as sfilter, _
q = cars_sql >> sfilter(_.cyl.isin([4, 6]) >> show_query()
```

**Comment by machow** [comment]
For the sqlalchemy error, it seems like maybe a bug in sqlalchemy-redshift? Here's code to reproduce what siuba's LazyTbl is running on init, when the error is raised.

```python
import sqlalchemy

connect_string = f'redshift+psycopg2://{db_user}:{db_password}@{db_host}:{db_port}/{db_name}'
source = sqlalchemy.create_engine(connect_string)
table_name = "sometable"

# NOTE: I think somedatabase in the example is being read as the schema (but could be wrong)
schema = "somedatabase" 

columns = []

sqlalchemy.Table(
                table_name,
                sqlalchemy.MetaData(bind = source),
                *columns,
                schema = schema,
                # true if no columns specified; will need to get them via reflection
                autoload_with = source if not columns else None
                )
```

Do you mind trying to create the table directly using the code above, and let me know if it works as expected? I don't have quick access to a redshift database anymore, but will set one up to be able to test these kinds of things!

**Comment by ofajardo** [comment]
hi thanks a lot for the quick reply!

Regarding the "in" thing, it makes fully sense and yes that helps, so I have upgraded successfully to siuba 0.0.20. I am aware of pandas "isin" and when I saw the "in" working in 0.0.17 I tought, wow, so nice! and got very confused by the error in 0.0.20, particularly because it doesn't say anything about the "in" clause, maybe the error message can include something about that. Because the temptation to use python's in as a SQL in is really big, I would suggest to put an example or a not about this in a more prominent place in the documentation, like when talking about databases, and maybe somewhere else too. 

Regarding the sqlalchemy thing, running the code you suggest raises exactly the same error as the one I saw with siuba, so that's the root of the issue. For now I have downgraded to sqlalchemy 1.3.15. Is this something that should be reported to sqlalchemy?

**Comment by ofajardo** [comment]
it seems that the issue is reported [here](https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/issues/195)

**Comment by machow** [comment]
Ah, helpful to see--thanks for chasing down the issue!

> and got very confused by the error in 0.0.20, particularly because it doesn't say anything about the "in" clause, maybe the error message can include something about that. 

This makes a lot of sense--I'll condense the caveats for using _ and put them in the docstring, and update the error message.

>  the temptation to use python's in as a SQL in is really big

I've been thinking about adding a siuba for SQL users doc. If you hit any other things that would be useful to tell SQL users up front, would love to hear them!

**Comment by ofajardo** [comment]
Thanks! I consider this solved so I close it.

**Comment by ofajardo** [comment]
the issue with slqlachemy-redshift is solved in its latetst version 0.7.9
